test/shared/utilities.ts

```diff
@@ -79,21 +80,51 @@ export async function getApprovalDigest(
   )
 }
 
-export async function mineBlock(provider: Web3Provider, timestamp: number): Promise<void> {
-  await new Promise(async (resolve, reject) => {
-    ;(provider._web3Provider.sendAsync as any)(
-      { jsonrpc: '2.0', method: 'evm_mine', params: [timestamp] },
-      (error: any, result: any): void => {
-        if (error) {
-          reject(error)
-        } else {
-          resolve(result)
-        }
-      }
-    )
-  })
+export function encodePrice(reserve0: BigNumber, reserve1: BigNumber) {
+  return [reserve1.mul(BigNumber.from(2).pow(112)).div(reserve0), reserve0.mul(BigNumber.from(2).pow(112)).div(reserve1)]
 }
 
-export function encodePrice(reserve0: BigNumber, reserve1: BigNumber) {
-  return [reserve1.mul(bigNumberify(2).pow(112)).div(reserve0), reserve0.mul(bigNumberify(2).pow(112)).div(reserve1)]
+export function sqrt(value: BigNumber) {
+  const ONE = BigNumber.from(1);
+  const TWO = BigNumber.from(2);
+  const x = BigNumber.from(value);
+  let z = x.add(ONE).div(TWO);
+  let y = x;
+  while (z.sub(y).isNegative()) {
+    y = z;
+    z = x.div(z).add(z).div(TWO);
+  }
+  return y;
 }
+
+// Calculate how much will be payed from liquidity as protocol fee in the next mint/burn
+export async function calcProtocolFee(_pair: DXswapPair, _factory: DXswapFactory) {
+  const [token0Reserve, token1Reserve, _] = await _pair.getReserves()
+  const kLast = await _pair.kLast()
+  const feeTo = await _factory.feeTo()
+  const protocolFeeDenominator = await _factory.protocolFeeDenominator()
+  const totalSupply = await _pair.totalSupply()
+  let rootK: BigNumber
+  let rootKLast: BigNumber
+  if (feeTo != AddressZero) {
+    // Check for math overflow when dealing with big big balances
+    const balance = sqrt((token0Reserve).mul(token1Reserve))
+    const balanceBN = balance ? bignumber(balance.toString()) : bignumber(0)
```

Comment:
Curious what `bignumber` can do that `BigNumber` (from ethers) can't do.

Reply:
`balanceBN` needs to be type `math.BigNumber` otherwise usage `log10` below is not possible.

Reply:
Didn't know that, thank you!